### PR TITLE
Add BatchDeploy

### PR DESCRIPTION
**Pull request by AbdelrahmanElawady**
### Description

Add the option to deploy batch deployments

### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-go/issues/116

### Checklist

- [ ] Tests included
- [x] Build pass
- [x] Documentation
- [x] Code format and docstring

### CODE REVIEW

File: grid-client/deployer/deployer.go [hunk 1]
```diff
@@ -383,6 +390,129 @@ func (d *Deployer) Wait(
...
+
+	mu := sync.Mutex{}
+
+	group, ctx2 := errgroup.WithContext(ctx)
+	for node, dls := range deployments {
+		client, err := d.ncPool.GetNodeClient(d.substrateConn, node)
+		if err != nil {
+			return map[uint32][]gridtypes.Deployment{}, errors.Wrap(err, "failed to get node client")
+		}
+		nodeClients[node] = client
```
Comment (xmonader) [inline grid-client/deployer/deployer.go]: isn't that client is already cached using the node connection pool?

File: grid-client/deployer/deployer.go [hunk 2]
```diff
@@ -383,6 +390,129 @@ func (d *Deployer) Wait(
...
+		for i, dl := range dls {
+			i := i
+			dl := dl
```
Comment (xmonader) [inline grid-client/deployer/deployer.go:409]: nicely done

File: grid-client/deployer/deployer.go [hunk 3]
```diff
@@ -383,6 +390,129 @@ func (d *Deployer) Wait(
...
+// BatchDeploy deploys a batch of deployments
+func (d *Deployer) BatchDeploy(ctx context.Context, deployments map[uint32][]gridtypes.Deployment, deploymentsSolutionProvider map[uint32][]*uint64) (map[uint32][]gridtypes.Deployment, error) {
+	deploymentsSlice := make([]gridtypes.Deployment, 0)
+	nodes := make([]uint32, 0)
+	contractsData := make([]substrate.BatchCreateContractData, 0)
+	nodeClients := make(map[uint32]*client.NodeClient)
+
+	mu := sync.Mutex{}
+
+	group, ctx2 := errgroup.WithContext(ctx)
```
Comment (xmonader) [inline grid-client/deployer/deployer.go:400]: ctx2 -> ctx

File: grid-client/deployer/deployer.go [hunk 4]
```diff
@@ -383,6 +390,129 @@ func (d *Deployer) Wait(
...
+		for i, dl := range dls {
+			i := i
+			dl := dl
+			node := node
+
+			group.Go(func() error {
+				select {
+				case <-ctx2.Done():
+					return nil
+				default:
```
Comment (xmonader) [inline grid-client/deployer/deployer.go:416]: excellent to do that one here for a non-blocking!

**Review by xmonader — APPROVED** [review]

File: grid-client/deployer/deployer.go [hunk 5]
```diff
@@ -383,6 +390,129 @@ func (d *Deployer) Wait(
 	return deploymentError
 }
 
+// BatchDeploy deploys a batch of deployments
+func (d *Deployer) BatchDeploy(ctx context.Context, deployments map[uint32][]gridtypes.Deployment, deploymentsSolutionProvider map[uint32][]*uint64) (map[uint32][]gridtypes.Deployment, error) {
+	deploymentsSlice := make([]gridtypes.Deployment, 0)
+	nodes := make([]uint32, 0)
```
Comment (rawdaGastan) [inline grid-client/deployer/deployer.go]: can we use nodes from `contractsData`?

File: grid-client/deployer/deployer.go (hunk 3 above)
Comment (AbdelrahmanElawady) [reply grid-client/deployer/deployer.go:400]: `group.Wait` cancels the context so other functions using ctx will immediately stop

**Review by rawdaGastan — APPROVED** [review]

**Review by rawdaGastan — APPROVED** [review]

**Pull request merged**